core/workflow/safestate_test.go

```diff
@@ -0,0 +1,158 @@
+			})
+		})
+		When("the role has one sub-role, the state is ERROR and is merged with MIXED state", func() {
+			It("should become MIXED", func() {
+				subroles := make([]Role, 1)
+				subroles[0] = &aggregatorRole{roleBase: roleBase{state: SafeState{state: task.MIXED}}}
+				ar := &aggregatorRole{
+					roleBase:   roleBase{state: SafeState{state: task.ERROR}},
+					aggregator: aggregator{subroles},
+				}
+				ar.state.merge(task.MIXED, ar)
+				Expect(ar.GetState()).To(Equal(task.MIXED))
+			})
+		})
+		When("the role has at least two sub-roles, state is ERROR and is merged with a healthy state", func() {
+			It("move to the corresponding healthy state", func() {
+				subroles := make([]Role, 2)
+				subroles[0] = &taskRole{
+					roleBase: roleBase{state: SafeState{state: task.CONFIGURED}},
+					Traits:   task.Traits{Trigger: "", Await: "", Timeout: "", Critical: true},
+				}
+				subroles[1] = &taskRole{
+					roleBase: roleBase{state: SafeState{state: task.CONFIGURED}},
+					Traits:   task.Traits{Trigger: "", Await: "", Timeout: "", Critical: true},
+				}
+				ar := &aggregatorRole{
+					roleBase:   roleBase{state: SafeState{state: task.ERROR}},
+					aggregator: aggregator{subroles},
+				}
+				ar.state.merge(task.CONFIGURED, ar)
+				Expect(ar.GetState()).To(Equal(task.CONFIGURED))
+			})
+		})
+		When("the role has at least two sub-roles, state is ERROR and is merged with a healthy state, but other sub-role stays in ERROR", func() {
+			It("should stay in ERROR", func() {
+				subroles := make([]Role, 2)
+				subroles[0] = &taskRole{
+					roleBase: roleBase{state: SafeState{state: task.ERROR}},
+					Traits:   task.Traits{Trigger: "", Await: "", Timeout: "", Critical: true},
+				}
+				subroles[1] = &taskRole{
+					roleBase: roleBase{state: SafeState{state: task.CONFIGURED}},
+					Traits:   task.Traits{Trigger: "", Await: "", Timeout: "", Critical: true},
+				}
+				ar := &aggregatorRole{
+					roleBase:   roleBase{state: SafeState{state: task.ERROR}},
+					aggregator: aggregator{subroles},
+				}
+				ar.state.merge(task.STANDBY, ar)
+				Expect(ar.GetState()).To(Equal(task.ERROR))
+			})
+		})
+		When("the role has at least two sub-roles, the state is healthy and one of non-critical sub-roles enters ERROR", func() {
+			It("should remain in the same state", func() {
+				subroles := make([]Role, 2)
+				subroles[0] = &taskRole{
+					roleBase: roleBase{state: SafeState{state: task.ERROR}},
+					Traits:   task.Traits{Trigger: "", Await: "", Timeout: "", Critical: false},
+				}
+				subroles[1] = &taskRole{
+					roleBase: roleBase{state: SafeState{state: task.CONFIGURED}},
+					Traits:   task.Traits{Trigger: "", Await: "", Timeout: "", Critical: true},
+				}
+				ar := &aggregatorRole{
+					roleBase:   roleBase{state: SafeState{state: task.CONFIGURED}},
+					aggregator: aggregator{subroles},
+				}
+				ar.state.merge(task.ERROR, ar)
+				Expect(ar.GetState()).To(Equal(task.CONFIGURED)) // fixme: this fails
```

Comment:
This seems to be related to https://its.cern.ch/jira/browse/OCTRL-853

Reply:
@justonedev1 ↑